src/LinearSolvers/preconditioners.jl

```diff
@@ -288,16 +288,16 @@ Update the preconditioner `p` from the sparse Jacobian `J` in CSC format for the
 Note that this implements the same algorithm as for the GPU and becomes very slow on CPU with growing number of blocks.
 
 """
-function update(J::SparseMatrixCSC, p)
-    kernel! = update_cpu_kernel!(CPU())
+function update(p, J::SparseMatrixCSC, device)
```

Comment:
The ordering of the argument is better, indeed. I would move one step further, and suggest the signature:
```julia
function update!(p, J::SparseMatrixCSC)
```
and remove the device in the argument

Reply:
I think we should use device consistently throughout the code. I think it shouldn't be decided based on the datatype. There is not safe map from datatype to device type. Oceananigans uses `arch` throughout the code. I like `device` better.